core/state_transition.go

```diff
@@ -245,127 +245,55 @@ func (st *StateTransition) to() common.Address {
-	if err != nil {
-		return nil, fmt.Errorf("parseSoulBalanceResp Unpack failed:%w", err)
-	}
+var (
+	slotArgs abi.Arguments
+)
 
-	returnValue := new(big.Int)
-	err = method.Outputs.Copy(&returnValue, returnValueInterface)
-	if err != nil {
-		return nil, fmt.Errorf("parseSoulBalanceResp Copy failed:%w", err)
-	}
-	returnValueU256, overflow := uint256.FromBig(returnValue)
-	if overflow {
-		return nil, fmt.Errorf("parsed soul balance overflow:%v", returnValue)
-	}
-	return returnValueU256, nil
+func init() {
+	uint64Ty, _ := abi.NewType("uint64", "", nil)
+	addressTy, _ := abi.NewType("address", "", nil)
+	slotArgs = abi.Arguments{{Name: "addr", Type: addressTy, Indexed: false}, {Name: "slot", Type: uint64Ty, Indexed: false}}
 }
 
-func burnSoulBalanceData(account common.Address, amount *big.Int) []byte {
-
-	method, ok := SoulGasTokenABI.Methods["burnFrom"]
-	if !ok {
-		panic("burnFrom method not found")
-	}
-
-	argument, err := method.Inputs.Pack(account, amount)
-	if err != nil {
-		panic("failed to pack argument")
-	}
-
-	data := make([]byte, len(method.ID)+len(argument))
-	copy(data[0:len(method.ID)], method.ID)
-	copy(data[len(method.ID):], argument)
-	return data
+func targetSlot(account common.Address) (slot common.Hash) {
+	data, _ := slotArgs.Pack(account, balancesSlot)
+	slot = crypto.Keccak256Hash(data)
+	return
 }
 
-func mintSoulBalanceData(account common.Address, amount *big.Int) []byte {
-	method, ok := SoulGasTokenABI.Methods["batchMint"]
-	if !ok {
-		panic("batchMint method not found")
-	}
-
-	argument, err := method.Inputs.Pack([]common.Address{account}, []*big.Int{amount})
-	if err != nil {
-		panic("failed to pack argument")
-	}
-
-	data := make([]byte, len(method.ID)+len(argument))
-	copy(data[0:len(method.ID)], method.ID)
-	copy(data[len(method.ID):], argument)
-	return data
+func (st *StateTransition) GetSoulBalance(account common.Address) *uint256.Int {
+	slot := targetSlot(account)
+	value := st.state.GetState(types.SoulGasTokenAddr, slot)
+	balance := new(uint256.Int)
+	balance.SetBytes(value[:])
+	return balance
 }
 
-var (
-	callSoulGasLimit = uint64(8_000_000)
-	// use hardcoded DEPOSITOR_ACCOUNT both as minter and burner
-	depositorAddress = common.HexToAddress("0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001")
-)
-
-func (st *StateTransition) GetSoulBalance(account common.Address) (*uint256.Int, error) {
-	// this evm call is free of gas charging
-	ret, _, vmerr := st.evm.StaticCall(vm.AccountRef(account), types.SoulGasTokenAddr, getSoulBalanceData(account), callSoulGasLimit)
-	if vmerr != nil {
-		return nil, vmerr
+func (st *StateTransition) SubSoulBalance(account common.Address, amount *big.Int) (err error) {
+	current := st.GetSoulBalance(account).ToBig()
+	if current.Cmp(amount) < 0 {
+		return fmt.Errorf("soul balance not enough, current:%v, expect:%v", current, amount)
 	}
-	return parseSoulBalanceResp(ret)
-}
+	var value common.Hash
+	current.Sub(current, amount).FillBytes(value[:])
```

Comment:
Can we use `uint256.MustFromBig(current.Sub(current, amount)).Bytes32()`?  This will address the uncertainty of the endianness issue.

Reply:
Great suggestion! Done, and the [test](https://github.com/blockchaindevsh/optimism/tree/TestModBalance) also changed to `Bytes32()`.